### PR TITLE
Deprecate ready promises

### PR DESCRIPTION
**Pull request by ggetz**
Will fix https://github.com/CesiumGS/cesium/issues/10909 by deprecating `readyPromise` pattern throughout the API.

### Why

The `readyPromise` pattern has been a problematic part of the CesiumJS API that does not match modern promise best practices. 
 * The background asynchronous nature can cause race conditions, especially if they do not properly handle destroyed objects. This has been coming up often in tests, leading to  "random" CI failures as mentioned in #10909. 
 * The existing API can [cause confusion around how to handle errors](https://community.cesium.com/t/cesium3dtileset-not-loading-tileset-json-from-any-url-and-silently-failing-with-bad-urls/16287). 
 * When `readyPromises` are unhandled, in many modern systems, a failed promise can lead to a critical error.

### How

The plan is to deprecate `readyPromise` properties throughout the API starting with the next release (1.103). As this affects high-touch areas of the API like imagery providers, terrain providers, and 3D Tiles, we'll remove the old code paths after three releases. 

The new implementation moves asynchronous operations out of synchronous constructors and into async factory functions such as `fromUrl`, which perform operations like loading necessary metadata up front, returning a promise that resolves to the created 3D tileset or provider. The created object can be assumed fully ready, and [all properties can be accessed immediately](https://community.cesium.com/t/urltemplateimageryprovider-broken-after-upgrade-from-1-91-to-1-92/22245/5) (such as `Cesium3DTileset.boundingSphere`). 3D tilesets or providers will still make requests for new tiles as changes to the camera and scene occur and is handled internally, the success or failure of which are exposed via events.

A few examples of this are as follows:

##### Before
```js
const tileset = scene.primitives.add(
  new Cesium.Cesium3DTileset({
    url: Cesium.IonResource.fromAssetId(1240402),
  })
);

tileset.readyPromise
  .then(function (tileset) {
    viewer.zoomTo(
      tileset,
      new Cesium.HeadingPitchRange(
        0.5,
        -0.2,
        tileset.boundingSphere.radius * 4.0
      )
    );
  })
  .catch(function (error) {
    console.log(error);
  });
```

##### After
```js
try {
  const tileset = await Cesium3DTileset.fromUrl(Cesium.IonResource.fromAssetId(1240402));
  scene.primitives.add(tileset);
  viewer.zoomTo(
      tileset,
      new Cesium.HeadingPitchRange(
        0.5,
        -0.2,
        tileset.boundingSphere.radius * 4.0
      )
    );
} catch (error) {
  console.log(error)
}
```

Besides deprecating the `readyPromise` pattern throughout the API, this change includes:
* Generally, a move towards using `async`/`await` syntax for flatter and more readable code
* Allow top level await in Sandcastle examples
* Adds a new async version of the` toThrowDeveloperError` matcher for better async testing flow

An un-finalized list of remaining TODOs include:

- [x] Fix warning on last line in sandcastle examples
- [x] Pass over async code for Sandcastles (and do we want to make the invoked wrapper function async). In particular, do a pass for race conditions when doing things like switching providers. https://github.com/CesiumGS/cesium/pull/11033#issuecomment-1404610180
- [x] `depthTestAgainstTerrain` setting via BaseLayerPicker? https://github.com/CesiumGS/cesium/pull/11033#issuecomment-1400538785
- [x] Finalize deprecation versions and open issues
- [x] `fromUrl` / `Resource` naming
-  [ ] There's a few places notes "ready is deprecated. This is here for backwards compatibility" which can be removed when the rest of the deprecated API is. This should be noted in the follow up deprecation issue.
- [x] Open forum thread to discuss API changes
- [x] Pass over jsDoc code examples 
- [x] Pass over internal documentation, ie https://github.com/CesiumGS/cesium/tree/45531ec9fc07786e0a3bda10ec6ba89c5507f153/Documentation/Contributors/PerformanceTestingGuide#how-to-measure-tile-load-time
- [ ] Pass over website tutorial examples

Contains work reviewed separately in the following PRs:
- https://github.com/CesiumGS/cesium/pull/11033
- https://github.com/CesiumGS/cesium/pull/11060
- https://github.com/CesiumGS/cesium/pull/11078
- https://github.com/CesiumGS/cesium/pull/11173

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ggetz!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/no-ready-promises/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/main/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

**Comment by ggetz** [comment]
@jjhembd I've run the builds on this branch a few times to verify no intermittent failures are popping up in CI due to race conditions. I believe we're in the clear.

Besides what has already been reviewed in other PRs, I've:
- [added a helper function,`CesiumTerrainProvider.fromIonAssetId` to mirror `Cesium3DTileset.fromIonAssetId`](https://github.com/CesiumGS/cesium/pull/11059/commits/ba022f70d5dae3f01da24d33a0b9f3fc4099b3cb). I don't believe any `fromIonAssetId` helper should be needed for imagery providers as they are covered by `IonImageryProvider`.
- [fixed a warning in Sandcastle](https://github.com/CesiumGS/cesium/pull/11059/commits/2dd4240e5840cf1438c4a40a93bd9db1dcb793e4)

Would you mind doing a pass on those updates? After that, I think this PR should be ready.

File: packages/engine/Source/Core/CesiumTerrainProvider.js [hunk 1]
```diff
@@ -1196,6 +1197,45 @@ CesiumTerrainProvider.prototype.getLevelMaximumGeometricError = function (
...
 
+/**
+ * Creates a {@link TerrainProvider} from a Cesium ion asset ID that accesses terrain data in a Cesium terrain format
+ * Terrain formats can be one of the following:
+ * <ul>
+ * <li> {@link https://github.com/AnalyticalGraphicsInc/quantized-mesh Quantized Mesh} </li>
+ * <li> {@link https://github.com/AnalyticalGraphicsInc/cesium/wiki/heightmap-1.0 Height Map} </li>
+ * </ul>
+ *
+ * @param {Resource|String|Promise<Resource>|Promise<String>} url The URL of the Cesium terrain server.
```
Comment (jjhembd) [inline packages/engine/Source/Core/CesiumTerrainProvider.js:1208]: This should be documented as an ID instead of a URL

**Comment by ggetz** [comment]
Thanks @jjhembd! Updated.

Additionally, considering there was a good amount of refactoring in gltf and 3D Tiles loading, I [verified loading performance was comparable](https://github.com/CesiumGS/cesium/tree/main/Documentation/Contributors/PerformanceTestingGuide). On average, time to load without cache was <2% faster on average in this branch as compared to `main`, and from cache <4% faster on average. I think it's safe to say these variances can be chalked up to error and that performance is unaffected.

**Comment by jjhembd** [comment]
I can confirm from my own tests: loading performance on this branch is comparable to main. This PR is looking good!

**Pull request merged**